addons/sale/models/sale_order.py

```diff
@@ -906,23 +915,50 @@ def _action_confirm(self):
                     order._create_analytic_account()
 
     def _send_order_confirmation_mail(self):
+        """ Send a mail to the customer each time an order is confirmed.
+
+        :return: None
+        """
+
+        for sale_order in self:
+            mail_template = sale_order._get_confirmation_template()
+            sale_order._send_order_notification_mail(mail_template)
+
+    def _send_payment_succeeded_for_order_mail(self):
+        """ Send a mail to the customer each time a payment is executed.
+
+        :return: None
+        """
+        for sale_order in self:
+            mail_template = self.env.ref(
+                'sale.mail_template_sale_payment_executed', raise_if_not_found=False
+            )
+            sale_order._send_order_notification_mail(mail_template)
+
+    def _send_order_notification_mail(self, mail_template):
+        """ Send a mail to the customer
+
+        Note: self.ensure_one()
+
+        :param mail.template mail_template: the template used to generate the mail
+        :return: None
+        """
         if not self:
             return
```

Comment:
```suggestion
        self.ensure_one()
        
        if not mail_template:
            return
```

In fact, this could get rid of the last if easily =] Sorry, should have seen that sooner

Reply:
yep, whenever called self is not empty :)